### PR TITLE
[SYCLomatic] Enable the migration of kernel function pointer 

### PR DESCRIPTION
**Pull request by intwanghao**

### CODE REVIEW

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 1]
```diff
@@ -5310,6 +5394,64 @@ DeviceFunctionInfo::DeviceFunctionInfo(size_t ParamsNum,
       FunctionName(FunctionName), IsLambda(false) {
   ParametersProps.resize(ParamsNum);
 }
+
+void DeviceFunctionInfo::collectInfoForWrapper(const FunctionDecl *FD) {
+  if (!WrapperInfoCollected) {
```
Comment (ziranzha) [inline clang/lib/DPCT/AnalysisInfo.cpp]: Move to DeviceFunctionDecl::collectInfoForWrapper.
Only generate kernelwrapper for decl that has body.

File: clang/lib/DPCT/AnalysisInfo.cpp [hunk 2]
```diff
@@ -5186,84 +5201,176 @@ class KernelPrinter {
...
+    // 1.Generate wrapper signature
+    if (ModuleUsed) {
+      Printer << "DPCT_EXPORT void " << FuncName << "_wrapper("
+              << MapNames::getClNamespace() << "queue &queue, const "
+              << MapNames::getClNamespace()
+              << "nd_range<3> &nr, unsigned int localMemSize, void "
+                 "**kernelParams, void **extra)";
+    } else {
+      if (!TParamsInfo.empty()) {
+        bool IsFirst = true;
```
Comment (ziranzha) [inline clang/lib/DPCT/AnalysisInfo.cpp]: May using `i==0` is better.

File: clang/lib/DPCT/AnalysisInfo.h [hunk 3]
```diff
@@ -2854,8 +2854,10 @@ class KernelCallExpr : public CallFunctionExpr {
       const std::pair<clang::tooling::UnifiedPath, unsigned> &LocInfo,
       const CallExpr *, bool IsAssigned = false);
   static std::shared_ptr<KernelCallExpr>
-  buildForWrapper(clang::tooling::UnifiedPath, const FunctionDecl *,
-                  std::shared_ptr<DeviceFunctionInfo>);
+  buildForWrapper(clang::tooling::UnifiedPath, const FunctionDecl *);
+  void setTemplateArgsStrForWrapper(std::string Str) {
+    TemplateArgsStrForWrapper = Str;
```
Comment (ziranzha) [inline clang/lib/DPCT/AnalysisInfo.h]:
```suggestion
    TemplateArgsStrForWrapper = std::move(Str);
```

File: clang/lib/DPCT/RulesLang/RulesLang.cpp [hunk 4]
```diff
@@ -4433,6 +4433,115 @@ void StreamAPICallRule::runRule(const MatchFinder::MatchResult &Result) {
...
+                this);
+  MF.addMatcher(unresolvedLookupExpr(unless(hasAncestor(cudaKernelCallExpr())))
+                    .bind("unresolvedRef"),
+                this);
+}
+
+void KernelCallRefRule::runRule(
+    const ast_matchers::MatchFinder::MatchResult &Result) {
+  if (auto DRE = getAssistNodeAsType<DeclRefExpr>(Result, "kernelRef")) {
+    if (auto ParentCE = dpct::DpctGlobalInfo::findAncestor<CallExpr>(DRE)) {
```
Comment (ziranzha) [inline clang/lib/DPCT/RulesLang/RulesLang.cpp]: Check the parent function decl is template instantiated, if so return directly.

File: clang/lib/DPCT/RulesLang/RulesLang.cpp [hunk 5]
```diff
@@ -4433,6 +4433,115 @@ void StreamAPICallRule::runRule(const MatchFinder::MatchResult &Result) {
...
+      if (FD && FD->hasAttr<CUDAGlobalAttr>()) {
+        KernelRefFound = true;
+        break;
+      }
+    }
+    if (!KernelRefFound) {
+      return;
+    }
+    if (auto ParentCE = dpct::DpctGlobalInfo::findAncestor<CallExpr>(ULE)) {
+      if (auto Callee = ParentCE->getDirectCallee()) {
```
Comment (ziranzha) [inline clang/lib/DPCT/RulesLang/RulesLang.cpp:4588]: If `Callee` is unresolved, may cause issue.

File: clang/lib/DPCT/RulesLang/RulesLang.cpp [hunk 6]
```diff
@@ -4536,24 +4652,61 @@ void KernelCallRule::runRule(
...
+              }
+            }
+          }
+        }
+        size_t ArgsNum = KCall->getNumArgs();
+        for (size_t i = 0; i < ArgsNum; i++) {
+          if (auto Arg = KCall->getArg(i)) {
+            if (!Arg->isDefaultArgument()) {
+              OS << ", " << ExprAnalysis::ref(Arg);
+            }
```
Comment (ziranzha) [inline clang/lib/DPCT/RulesLang/RulesLang.cpp:4737]:
```suggestion
            } else {
              break;
            }
```

File: clang/lib/DPCT/RulesLang/RulesLang.cpp [hunk 7]
```diff
@@ -4620,8 +4773,51 @@ void KernelCallRule::runRule(
...
+    bool IsFuncTypeErased = true;
+    auto QT = CalleeDRE->getType();
+    if (QT->isFunctionType()) {
+      IsFuncTypeErased = false;
+    } else if (QT->isPointerType()) {
+      const Type *PointeeType = QT->getPointeeType().getTypePtr();
+      if (PointeeType->isFunctionType()) {
+        IsFuncTypeErased = false;
+      }
+    }
```
Comment (ziranzha) [inline clang/lib/DPCT/RulesLang/RulesLang.cpp]: Can refine the code by try getting pointee type at first.

File: clang/lib/DPCT/RulesLang/RulesLang.cpp [hunk 8]
```diff
@@ -4620,8 +4773,51 @@ void KernelCallRule::runRule(
...
+    if (IsFuncTypeErased) {
+      DpctGlobalInfo::setCVersionCUDALaunchUsed();
+    }
```
Comment (ziranzha) [inline clang/lib/DPCT/RulesLang/RulesLang.cpp]: Move into the next if statement.

File: clang/lib/DPCT/Utility.cpp [hunk 9]
```diff
@@ -4998,21 +4998,57 @@ int isArgumentInitialized(
 
   return DeclsRequireInit.empty();
 }
-const DeclRefExpr *getAddressedRef(const Expr *E) {
+const Expr *getAddressedRef(const Expr *E, const FunctionDecl **FuncDecl,
+                            bool IsCheckFunctionDecl) {
```
Comment (ziranzha) [inline clang/lib/DPCT/Utility.cpp]:
```suggestion
const Expr *getAddressedRef(const Expr *E, bool IsCheckFunctionDecl = false,
                            const FunctionDecl **FuncDecl = nullptr) {
```

File: clang/runtime/dpct-rt/include/dpct/kernel.hpp [hunk 10]
```diff
@@ -444,6 +444,85 @@ static inline void invoke_kernel_function(dpct::kernel_function &function,
            localMemSize, kernelParams, extra);
 }
 
+class kernel_launch {
+  template <typename FuncT, typename ArgSelector, std::size_t... Index>
+  static void launch_helper(FuncT *func, ArgSelector &selector,
```
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/dpct/kernel.hpp]:
```suggestion
  static void launch_helper(FuncT &&func, ArgSelector &selector,
```

File: clang/runtime/dpct-rt/include/dpct/kernel.hpp [hunk 11]
```diff
@@ -444,6 +444,85 @@ static inline void invoke_kernel_function(dpct::kernel_function &function,
...
+        static_cast<sycl::range<3>>(group_range * local_range),
+        static_cast<sycl::range<3>>(local_range));
+    _local_mem_size = local_mem_size;
+  };
+
+public:
+  static inline thread_local sycl::queue *_que = nullptr;
+  static inline thread_local sycl::nd_range<3> _nr = sycl::nd_range<3>();
+  static inline thread_local unsigned int _local_mem_size = 0;
+  static inline thread_local std::map<
```
Comment (ziranzha) [inline clang/runtime/dpct-rt/include/dpct/kernel.hpp]:
```suggestion
  static inline std::map<
```

File: clang/test/dpct/launch-kernel-cooperative.cu [hunk 12]
```diff
@@ -78,17 +78,10 @@ int main() {
   // CHECK-NEXT:       });
   // CHECK-NEXT:   });
   cudaLaunchCooperativeKernel((const void *)&template_kernel<int>, dim3(16), dim3(16), args, 32, stream);
-
+  // CHECK: void *kernel_func = (void *)&dpct::wrapper_register(kernel_wrapper).get();
```
Comment (ziranzha) [inline clang/test/dpct/launch-kernel-cooperative.cu]:
```suggestion
  // CHECK: void *kernel_func = (void *)dpct::wrapper_register(&kernel_wrapper);
```

File: clang/lib/DPCT/RulesLang/RulesLang.cpp [hunk 13]
```diff
@@ -4442,38 +4442,112 @@ void KernelCallRefRule::registerMatcher(ast_matchers::MatchFinder &MF) {
...
-      auto &Map = DpctGlobalInfo::getWrapperRegisterMap();
-      auto Key = getStrFromLoc(SM.getSpellingLoc(DRE->getBeginLoc()));
-      if (!Map.count(Key)) {
-        Map.insert({getStrFromLoc(SM.getSpellingLoc(DRE->getBeginLoc())),
-                    {InsertBeforeStmt(DRE, MapNames::getDpctNamespace() +
-                                               "wrapper_register(")
-                         .getReplacement(DpctGlobalInfo::getContext()),
-                     InsertAfterStmt(DRE, ").get()")
-                         .getReplacement(DpctGlobalInfo::getContext())}});
+    std::cout << IsTemplateRelated << std::endl;
```
Comment (ziranzha) [inline clang/lib/DPCT/RulesLang/RulesLang.cpp]: Remove debug code.

File: clang/lib/DPCT/RulesLang/RulesLang.cpp [hunk 14]
```diff
@@ -4442,38 +4442,112 @@ void KernelCallRefRule::registerMatcher(ast_matchers::MatchFinder &MF) {
...
+      if ((OuterFD->getTemplatedKind() ==
+           FunctionDecl::TemplatedKind::TK_NonTemplate) ||
+          (OuterFD->getTemplatedKind() ==
+           FunctionDecl::TemplatedKind::TK_FunctionTemplate)) {
+        std::string TypeRepl;
+        if (DpctGlobalInfo::isCVersionCUDALaunchUsed() && IsTemplateRelated &&
+            !DRE->hasExplicitTemplateArgs()) {
+          TypeRepl = getTypeRepl(DRE);
+        }
+        std::cout << TypeRepl << std::endl;
```
Comment (ziranzha) [inline clang/lib/DPCT/RulesLang/RulesLang.cpp]: Remove debug code.

**Review by ziranzha — APPROVED** [review]

File: clang/lib/DPCT/RulesLang/RulesLang.h [hunk 15]
```diff
@@ -442,13 +442,24 @@ class KernelCallRule : public NamedMigrationRule<KernelCallRule> {
...
-  removeTrailingSemicolon(const CallExpr *KCall,
-                          const ast_matchers::MatchFinder::MatchResult &Result);
+  SourceLocation findAndRemoveTrailingSemicolon(
+      const CallExpr *KCall,
+      const ast_matchers::MatchFinder::MatchResult &Result, bool Remove = true);
   void instrumentKernelLogsForCodePin(const CUDAKernelCallExpr *KCall,
                                       SourceLocation &EpilogLocation);
 };
 
+class KernelCallRefRule : public NamedMigrationRule<KernelCallRefRule> {
```
Comment (zhimingwang36) [inline clang/lib/DPCT/RulesLang/RulesLang.h:460]: Could you summary some comment on functionality of the new KernelCallRefRule?

File: clang/lib/DPCT/RulesLang/RulesLang.h [hunk 16]
```diff
@@ -442,13 +442,24 @@ class KernelCallRule : public NamedMigrationRule<KernelCallRule> {
...
+  std::string getTypeRepl(const Expr *E);
+  template <typename T>
+  void insertWrapperPostfix(const T *Node, std::string &&TypeRepl,
```
Comment (zhimingwang36) [inline clang/lib/DPCT/RulesLang/RulesLang.h:463]: add comment for the main function of this rule.

File: clang/runtime/dpct-rt/include/dpct/kernel.hpp [hunk 17]
```diff
@@ -444,6 +444,85 @@ static inline void invoke_kernel_function(dpct::kernel_function &function,
            localMemSize, kernelParams, extra);
 }
 
+class kernel_launch {
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/kernel.hpp]: add comment for class, static function, and publich function API

File: clang/runtime/dpct-rt/include/dpct/kernel.hpp [hunk 18]
```diff
@@ -444,6 +444,85 @@ static inline void invoke_kernel_function(dpct::kernel_function &function,
...
+public:
+  static inline thread_local sycl::queue *_que = nullptr;
+  static inline thread_local sycl::nd_range<3> _nr = sycl::nd_range<3>();
+  static inline thread_local unsigned int _local_mem_size = 0;
+  static inline std::map<
+      const void *,
+      std::function<void(dim3, dim3, void **, unsigned int, queue_ptr)>>
+      wrapper_map = {};
+
+  static void regifter_kernel_launcher(
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/kernel.hpp]: ==> register_kernel_ptr

File: clang/runtime/dpct-rt/include/dpct/kernel.hpp (hunk 17 above)
Comment (zhimingwang36) [reply clang/runtime/dpct-rt/include/dpct/kernel.hpp:447]: Maybe with example on how to use this class.

File: clang/runtime/dpct-rt/include/dpct/kernel.hpp [hunk 19]
```diff
@@ -444,6 +444,85 @@ static inline void invoke_kernel_function(dpct::kernel_function &function,
...
+      const void *func,
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/kernel.hpp]: May comment on when this this helper function.

File: clang/runtime/dpct-rt/include/dpct/util.hpp [hunk 20]
```diff
@@ -1185,6 +1187,13 @@ class args_selector<n_nondefault_params, n_default_params, R(Ts...)> {
   }
 };
 
+template <typename Func, std::size_t N> struct nth_argument_type {
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/util.hpp:1201]: What is the N mean here?  nth?  Please add comment for end user.

File: clang/test/dpct/function_pointer.cu [hunk 21]
```diff
@@ -0,0 +1,186 @@
...
+
+    delete[] h_A;
+    delete[] h_B;
+    delete[] h_C;
+}
+
+template <typename T>
+void hoo() {
+    // CHECK: fpt<int> a = dpct::wrapper_register<decltype(a)>(vectorTemplateAdd_wrapper);
+  fpt<int> a = vectorTemplateAdd;
```
Comment (zhimingwang36) [inline clang/test/dpct/function_pointer.cu:176]: a is declared,  but not used?

File: clang/test/dpct/launch-kernel-cooperative-usm.cu [hunk 22]
```diff
@@ -24,6 +24,23 @@ __global__ void kernel(int *d, cudaTextureObject_t tex) {
   tex1D(d + gtid, tex, gtid);
 }
 
+// CHECK: void kernel_wrapper(int * d ,dpct::image_wrapper_base_p tex) {
```
Comment (zhimingwang36) [inline clang/test/dpct/launch-kernel-cooperative-usm.cu:27]: for new generate kernel_wrapper,  let's add a comment line to tell
// Auto generated SYCL kernel wrarpper used to ....

File: clang/runtime/dpct-rt/include/dpct/kernel.hpp [hunk 23]
```diff
@@ -444,6 +444,85 @@ static inline void invoke_kernel_function(dpct::kernel_function &function,
...
+  };
+
+public:
+  static inline thread_local sycl::queue *_que = nullptr;
+  static inline thread_local sycl::nd_range<3> _nr = sycl::nd_range<3>();
+  static inline thread_local unsigned int _local_mem_size = 0;
+  static inline std::map<
+      const void *,
+      std::function<void(dim3, dim3, void **, unsigned int, queue_ptr)>>
+      wrapper_map = {};
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/kernel.hpp]: wrapper_map ==>  kernel_function_ptr_map  //get launchable function pointer from raw pointer

**Review by zhimingwang36 — APPROVED** [review]

**Pull request merged**